### ISSUE TITLE
Implement collision-resistant hashmap

### ISSUE THREAD

**Issue by Crypt-iQ** [google/AFL#65]
This is sort of a feature request, sort of me asking for tips on how to go about implementing this in the codebase.  After reading the CollAFL paper (https://chao.100871.net/papers/oakland18.pdf), it seems like a collision-resistant hashing when instrumenting is ideal for fuzzing, and also seems like low-hanging fruit.  When fuzzing large codebases, collisions are the norm.  From a usability perspective, this is bad since code paths won't be explored if they collide!

**Comment by vanhauser-thc** [comment]
its sadly not a low hanging fruit. the if there are many edges the only way of doing this collision free is to collect all instruction points (or parts of it), and that slows down the fuzzing process too much.

if there are not too many edges you can avoid most collisions by selecting the IDs for each edge instead of doing it random. This sounds easy but is actually hard. Its easily done in a function, but from a function callee to a function caller this is not possible to resolve so this can still collide. Also if there are too many edges it cant work. I implemented that in a branch here: https://github.com/vanhauser-thc/AFLplusplus/tree/non_colliding_edge_cov

You can also just increase the map (in config.h) but the increase does not help that much. And it slows fuzzing down - a lot.

So ... currently there is not a good solution. So better use a mix:
* use the AFL_LLVM_WHITELIST feature
* ues AFL_LLVM_INSTRIM mode, because it is less costly, faster, less instrumentation necessary.
* recompile the target for every instance so at least the collisions are different and you are able to identify paths that otherwise would always collide.

for AFL_LLVM_INSTRIM and AFL_LLVM_WHITELIST you need to use afl++ though.

**Comment by jonathanmetzman** [comment]
> its sadly not a low hanging fruit. the if there are many edges the only way of doing this collision free is to collect all instruction points (or parts of it), and that slows down the fuzzing process too much.
> 

Indeed. libFuzzer has this problem where for large targets, much of the time will be spend iterating over the coverage map.
